airbyte-integrations/connectors/source-netsuite-odbc/source_netsuite_odbc/streams.py

```diff
@@ -0,0 +1,424 @@
+      return primary_key_sorter
+    
+    # How does Incremental Refresh work?
+    # We first split things into stream slices.  Then, we start with filtering our SQL query by > the start date of the stream slice
+    # and <= the end date of the stream slice.  We then sort by the incremental column, and then by the primary key column (s).
+    #  As we fetch records, we start filtering by > the most recent incremental time we've seen. We also add on a filter for 
+    # = the most recent incremental time but with a primary key > the last one we've seen.  This ensures that we don't miss any records
+    # in the case where page size = N and there are N + 1 records with the same incremental time. 
+    def generate_incremental_query(self, stream_slice):
+      if self.incremental_column is None:
+        raise AirbyteTracedException.from_exception(message=GENERATING_INCREMENTAL_QUERY_WITH_NO_INCREMENTAL_COLUMN_ERROR, failure_type=FailureType.system_error)
+      values = ', '.join(self.properties) # we use values instead of '*' because we want to preserve the order of the columns
+      incremental_column_sorter = f"{self.incremental_column} ASC" if self.incremental_column else ""
+      incremental_filter = self.generate_incremental_filter_for_incremental_refresh(stream_slice)
+      primary_key_filter = self.generate_primary_key_filter_for_incremental_refresh(stream_slice)
+      primary_key_sorter = self.generate_primary_key_sorter()
+      query = f"""
+        SELECT TOP {NETSUITE_PAGINATION_INTERVAL} {values} FROM {self.table_name}
+        WHERE {incremental_filter} OR {primary_key_filter}
+        ORDER BY {incremental_column_sorter}, {primary_key_sorter}
+      """
+      return query
+    
+    def generate_primary_key_filter_for_incremental_refresh(self, stream_slice):
+      primary_key_filter = self.generate_primary_key_filter()
+      starting_timestamp = self.incremental_most_recent_value_seen if self.incremental_most_recent_value_seen is not None else stream_slice['first_day']
+      timestamp_filter = ""
+      if starting_timestamp is not None:
+        starting_timestamp = parser.parse(str(starting_timestamp)).strftime('%Y-%m-%d %H:%M:%S.%f')
+        timestamp_filter = f"{self.incremental_column} = to_timestamp('{starting_timestamp}', 'YYYY-MM-DD HH24:MI:SS.FF')"
+      else:
+        timestamp_filter = f"{self.incremental_column} IS NULL"
+      return f"({timestamp_filter} AND {primary_key_filter})"
+    
+    def generate_incremental_filter_for_incremental_refresh(self, stream_slice):
+      if self.incremental_column is None:
+        raise AirbyteTracedException.from_exception(message=GENERATING_INCREMENTAL_QUERY_WITH_NO_INCREMENTAL_COLUMN_ERROR, failure_type=FailureType.system_error)
+      # first, we check to see if we've been given a null slice, and if so, we change the query accordingly
+      if (stream_slice['first_day'] is None or stream_slice['last_day'] is None):
```

Comment:
is this ever expected? If yes, can you add a comment explaining why? If not, we should raise an exception